### ISSUE TITLE
Deploy with tarball

### ISSUE THREAD

**Issue by hotrush** [banago/PHPloy#259]
Maybe will be good to have an option to deploy with tarball. Make an archive of changed files, upload to any temp place, extract and replace files. Because uploading long lists can be problematic for any ftp servers.

```
 ^  173 of 4134 vendor//monolog/monolog/src/Monolog/Handler/DeduplicationHandler.php
 ^  174 of 4134 vendor//monolog/monolog/src/Monolog/Handler/ZendMonitorHandler.php
 ^  175 of 4134 vendor//monolog/monolog/src/Monolog/Handler/BufferHandler.php
PHP Warning:  ftp_fput(): Entering Passive Mode (45,34,36,92,16,64) in phar:///home/rush/Projects/nest/phploy.phar/vendor/league/flysystem/src/Adapter/Ftp.php on line 211
 ! Failed to upload vendor//monolog/monolog/src/Monolog/Handler/AbstractSyslogHandler.php.
 ^  176 of 4134 vendor//monolog/monolog/src/Monolog/Handler/AbstractSyslogHandler.php
```

**Comment by ankitce** [comment]
I am also looking for the same feature. This will make the fresh deployment a breeze.

**Comment by banago** [comment]
@hotrush and @ankitce ths is a great proposal. if you would like to implement this feature, I'd be more then happy to merge. Thanks.

**Comment by EdwinHoksberg** [comment]
I'm willing to implement this feature, but how would unpacking an uploaded archive work on a remote server with just FTP access? 

@hotrush @ankitce

**Comment by banago** [comment]
@EdwinHoksberg thanks for that. Personally I don't know how that's going to work.

**Comment by hotrush** [comment]
@EdwinHoksberg Yes, only sftp can access shell. But some hacks possible, for example any uploaded temporary script available on any public url that will run shell script.

**Comment by EdwinHoksberg** [comment]
Yea I guess that would be the only way. What about just creating an archive and upload it and let the user handle the unpacking? That would stop us from creating any 'hackish' features.

What do you think about that @banago?

**Comment by ankitce** [comment]
Here is my proposal for development:
1) At the point where the files are being uploaded, intead of uploading the files add them to an archive in local tmp folder.
2) File delete statements should not be altered, the files (which are deleted in repo) should still be deleted one by one on server.
3) Upload the archive on server.
4) After successful upload of archive, upload a PHP utility file which can extract the archive contents. For example, if the PHP script filename is unzipHashArchive.php, accessing the script "wget DOMAIN_NAME/unzipHashArchive.php?archive=SOMEHASH_OR_FILENAME" should extract the file.
 Alternatively, SSH connection can also do the job.
5) In the post deploy action, delete the archive file and the PHP script (unzipHashArchive.php) used to extract the archive.

Please see if there is any improvement needed in the flow @EdwinHoksberg @hotrush @banago

**Comment by EdwinHoksberg** [comment]
That seems like a good process.
The only thing that might not work is calling the unzip script from a url, because you can't make sure that you are uploading files to a document root that the webserver is serving, or that there is a webserver running at all. 
Ofcourse this only counts with FTP(S), we can use SFTP to easily extract files.

**Comment by hotrush** [comment]
@EdwinHoksberg i think path to script and public uri must be configurable. webserver must allow execute files with our pre-configured name (not index.php).

**Comment by banago** [comment]
It feels very hackish for FTP to make this work. Let's implement this only for SFTP and degrade to normal upload for FTP, even if this options is selected.
